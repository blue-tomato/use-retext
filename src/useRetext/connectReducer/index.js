import { mapValuesDeep } from '../../helpers';

export default (reducer, emitter, onCall) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({ key, scope: reducerScope, reducer: value, payload });
    });
  });
};
