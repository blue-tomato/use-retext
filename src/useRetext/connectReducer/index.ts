import { Emitter } from 'mitt';
import { mapValuesDeep } from '../../helpers';
import { Reducer, ReducerFuntion } from '../../Store';

interface OnCallObject {
  key: string;
  scope: string;
  reducer: ReducerFuntion;
  payload: any;
}

export default (reducer: Reducer, emitter: Emitter, onCall: (event: OnCallObject) => void) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({ key, scope: reducerScope, reducer: value, payload });
    });
  });
};
