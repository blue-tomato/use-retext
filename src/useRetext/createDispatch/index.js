import mitt from 'mitt';
import { mapValuesDeep } from '../../helpers';

export default action => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => payload => {
    emitter.emit(key, { scope, payload });
  });

  return [dispatch, emitter];
};
