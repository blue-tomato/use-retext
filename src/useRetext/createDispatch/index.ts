import mitt from 'mitt';
import { mapValuesDeep } from '../../helpers';
import { Action } from '../../Store';

export default <S>(action: Action<S>) => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => (payload: any) => {
    emitter.emit(key, { scope, payload });
  });

  return [dispatch, emitter];
};
