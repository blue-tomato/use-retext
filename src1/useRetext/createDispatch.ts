import mitt from 'mitt';
import { mapValuesDeep } from '../helpers';

export default (action: object): [unknown, mitt.Emitter] => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => (payload?: unknown) =>
    emitter.emit(key, { scope, payload }),
  );

  return [dispatch, emitter];
};
