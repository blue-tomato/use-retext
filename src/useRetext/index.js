import { useMemo, useState } from 'react';
import { get } from 'lodash-es';
import { assert } from '../helpers';
import createDisptach from './createDispatch';
import connectReducer from './connectReducer';

const mapReducer = (reducer, setState) =>
  Object.entries(reducer).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]:
        typeof value === 'object' ? mapReducer(value, setState) : payload => setState(state => value(state, payload)),
    }),
    {},
  );

export default store => {
  assert(typeof store !== 'object', 'Store is not an object');

  const { state: initialState, action, reducer } = store;

  assert(
    typeof initialState !== 'object' || typeof action !== 'object' || typeof reducer !== 'object',
    'State, action or reducer are not an object',
  );

  const [state, setState] = useState(initialState);
  const [disptach, emitter] = createDisptach(action);

  connectReducer(reducer, emitter, ({ scope, reducer: reducerFunction, payload }) => {
    // reducerFunction(get(state, scope), payload);
  });

  const dispatch = useMemo(() => mapReducer(reducer, setState), [reducer]);

  return { state, dispatch };
};
