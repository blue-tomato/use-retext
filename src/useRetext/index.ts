import { useState } from 'react';
import get from 'just-safe-get';
import set from 'just-safe-set';
import { assert } from '../helpers';
import createDispatch from './createDispatch';
import connectReducer from './connectReducer';
import { Store } from '../Store';

export default (store: Store) => {
  assert(typeof store !== 'object', 'Store is not an object');

  const { state: initialState, action, reducer } = store;

  assert(
    typeof initialState !== 'object' || typeof action !== 'object' || typeof reducer !== 'object',
    'State, action or reducer are not an object',
  );

  const [state, setState] = useState(initialState);
  const [dispatch, emitter] = createDispatch(action);

  connectReducer(reducer, emitter, ({ scope, reducer: reducerFunction, payload }) => {
    setState((currentState: any) => {
      const scopedState = scope ? get(currentState, scope) : currentState;
      const result = reducerFunction(scopedState, payload);
      let newState = { ...currentState };

      if (scope) {
        set(newState, scope, { ...scopedState, ...result });
      } else {
        newState = { ...scopedState, ...result };
      }

      return newState;
    });
  });

  return { state, dispatch };
};
