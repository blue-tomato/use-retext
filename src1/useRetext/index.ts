import { useState } from 'react';
import get from 'just-safe-get';
import set from 'just-safe-set';
import typeOf from 'just-typeof';
import { assert } from '../helpers';
import { Store } from '../Store';
import extractInternals from './extractInternals';
import createDispatch from './createDispatch';
import registerReducer from './registerReducer';

export type SelectorFunction = (store: Store) => Store;

export default (store: Store, selectorFunction?: SelectorFunction) => {
  assert(typeOf(store) !== 'object', 'Store is not an object');

  const selectedStore = selectorFunction ? selectorFunction(store) : store;
  const [initialState, actions, reducer] = extractInternals(selectedStore);
  const [state, setState] = useState(initialState);
  const [dispatch, emitter] = createDispatch(actions);

  registerReducer(reducer, emitter, ({ reducerFunction, scope, payload }) => {
    setState((currentState: object) => {
      const scopedState = (scope ? get(currentState, scope) : currentState) as object;
      const result = reducerFunction(scopedState, payload);
      let newState = { ...currentState };

      if (scope) {
        set(newState, scope, { ...scopedState, ...result });
      } else {
        newState = { ...scopedState, ...result };
      }

      return newState;
    });
  });

  return { state, dispatch };
};
