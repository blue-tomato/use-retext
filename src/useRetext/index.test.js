import { renderHook, act } from '@testing-library/react-hooks';
import useRetext from '.';

const getHook = store => renderHook(() => useRetext(store));
const { store } = global;

describe('store argument', () => {
  it('throws an error when no object is passed', () => {
    const { result } = getHook();

    expect(result.error).toMatchSnapshot();
  });

  it('throws an error when no state is passed', () => {
    const { result } = getHook({ reducer: {} });

    expect(result.error).toMatchSnapshot();
  });

  it('throws an error when no reducer is passed', () => {
    const { result } = getHook({ state: {} });

    expect(result.error).toMatchSnapshot();
  });
});

describe('return', () => {
  it('keeps the initial state', () => {
    const { result } = getHook(store);
    const [state] = result.current;

    expect(state).toMatchSnapshot();
  });

  it('creates a disptach object', () => {
    const { result } = getHook(store);
    const [, dispatch] = result.current;

    expect(dispatch).toMatchSnapshot();
  });
});

describe('in application usage', () => {
  it('tests the count', () => {
    const { result } = getHook(store);

    act(() => {
      result.current[1].setCount(3);
      result.current[1].increment();
      result.current[1].increment();
      result.current[1].decrement();
    });

    expect(result.current[0].count).toMatchSnapshot();
  });

  it('tests the sideMenu', () => {
    const { result } = getHook(store);

    act(() => {
      result.current[1].sideMenu.toggle();
    });

    expect(result.current[0].sideMenu.isOpen).toMatchSnapshot();
  });
});
