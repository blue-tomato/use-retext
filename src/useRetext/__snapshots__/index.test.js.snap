// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`in application usage tests the count 1`] = `4`;

exports[`in application usage tests the sideMenu 1`] = `true`;

exports[`return creates a disptach object 1`] = `
Object {
  "decrement": [Function],
  "increment": [Function],
  "setCount": [Function],
  "sideMenu": Object {
    "toggle": [Function],
  },
}
`;

exports[`return keeps the initial state 1`] = `
Object {
  "count": 0,
  "sideMenu": Object {
    "isOpen": false,
  },
}
`;

exports[`store argument throws an error when no object is passed 1`] = `[Error: Store is not an object]`;

exports[`store argument throws an error when no reducer is passed 1`] = `[Error: State or reducer are not an object]`;

exports[`store argument throws an error when no state is passed 1`] = `[Error: State or reducer are not an object]`;
